dal/dal.go

```diff
@@ -116,22 +120,101 @@ func (d *Dal) KeyExists(key string) (bool, bool, error) {
 	return true, resp.Node.Dir, nil
 }
 
-// An unwieldy wrapper for setting a new key
-func (d *Dal) Set(key, value string, dir bool, ttl int, prevExist string) error {
-	existState := client.PrevExistType(prevExist)
+type SetOptions struct {
+	// If SetOptions.Dir=true then value is ignored.
+	Dir       bool
+	TTLSec    int
+	PrevExist string
+
+	// Create parents will recursively create parent directories as needed.
+	CreateParents bool
+
+	// To be used with create parents.
+	// If depth > 0 it will only create depth number of parents.
+	// If depth < 0 it will try to create as many parents as necessary.
+	// If depth == 0 it will not create any parents. This behaves the same as Set().
+	Depth int
+}
+
+// A wrapper for setting a new key
+func (d *Dal) Set(key, value string, opt *SetOptions) error {
+	if key[0] != '/' {
+		key = "/" + key
+	}
+
+	// enforce this just in case
+	if !opt.CreateParents {
+		opt.Depth = 0
+	}
+
+	//TODO handle the case that depth is lager than the num of path elements
+
+	err := d.setAndCreateParents(
+		d.Prefix+key,
+		value,
+		opt.Dir,
+		opt.Depth,
+		time.Duration(opt.TTLSec)*time.Second,
+		client.PrevExistType(opt.PrevExist),
+	)
+
+	return err
+}
+
+// Helper method for set to recursively create its parents if they do not exist.
+// It will create up to `depth` number of parents.
+func (d *Dal) setAndCreateParents(
+	key, value string, dir bool, depth int, ttl time.Duration, pExist client.PrevExistType) error {
 
```

Comment:
I think we should get the depth check in before we merge this. I really dig the idea that we have 'control' over how deep we should create this, but we should definitely not be able to pass bogus crap to it.

I think a *very* simple way to this would be to just do a `strings.Split('/')` count the elements we've got and verify that the num elements is not larger than depth. This would trigger before we ever attempt a .Set() -- seems legit?